Improve PageProcessor retained bytes calculations

pettyjamesm opened the pull request:
## Description
Follows up from https://github.com/trinodb/trino/pull/25600 by account for all necessary components of `SourcePage` implementations in `SourcePage#getRetainedBytes()` and `SourcePage#retainedBytesForEachPart` methods. Also includes `SelectedPositions` and `Block[] previouslyComputedResults` in the `PageProcessor.ProjectSelectedPositions` retained size calculation.

<!-- Mark the appropriate option with an (x). Propose a release note if you can.
More info at https://trino.io/development/process#release-note -->
## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

wendigo commented on core/trino-spi/src/main/java/io/trino/spi/connector/SourcePage.java:
```diff
@@ -66,6 +66,12 @@ static SourcePage create(Page page)
+     * Returns the retained size of the {@link SourcePage} container and associated fields without including
+     * the size of any field blocks (loaded or unloaded).
+     */
+    long retainedShallowSizeInBytes();
```
nit: shallowRetainedSizeInBytes ;-)

pettyjamesm replied:
I'll wait for everyone to weigh in with their favorite alternative method name before I go update all of the commits to match 😛

pettyjamesm replied:
Renamed

wendigo commented:
@martint I think we should also land this before the release

raunaqmorarka commented:
> @martint I think we should also land this before the release

I don't think this is problematic enough to block the release

wendigo commented:
@raunaqmorarka we've been blocked for 2 weeks on that, 1 or 2 more days doesn't make a difference

dain reviewed (commented):
Seems fine to me.  My main question is can we avoid adding the new method if we update the `retainedBytesForEachPart` methods to report "this" size and sizes for internal block arrays and such.  This is how RLE block works for example.

pettyjamesm commented:
> My main question is can we avoid adding the new method if we update the `retainedBytesForEachPart` methods to report "this" size and sizes for internal block arrays and such. This is how RLE block works for example.

Yeah, that approach reduces the surface area of the `SourcePage` interface so I've updated the PR to do that instead. I've also added logic to include some other PageProcessor fields that previously weren't being considered in the retained size calculation.

dain reviewed (approved):
This is great
